src/modules/thresholdsig/threshold.md

```diff
@@ -0,0 +1,282 @@
+   coefficients, runs `secp256k1_musig_verify_shard` to check consistency
+   of these. This function outputs a modified set of public keys `{Z_i}`
+   and a modified secret key `z_i`, which the signer should use in place of
+   the keys produced in steps 2 and 3. (Specifically, on the first invocation
+   `secp256k1_musig_verify_shard` outputs keys which it updates on subsequent
+   invocations, until all participants' shards and coefficients have been
+   taken into account.)
+6. [TODO] She signs the each set of coefficients and broadcasts her signatures.
+7. [TODO] each participant verifies each others' signatures.
+
+Key Generation Flowchart:
+```
+Me:                  Others:
+
+ec_pubkey_create ->
+                  <- ec_pubkey_create
+pubkey_combine
+                     pubkey_combine
+
+if k < n {
+keysplit         ->
+                  <- keysplit
+verify_shard
+                     verify_shard
+sign_coefficient ->
+                  <- sign_coefficient
+verify_signature
+                     verify_signature
+}
+```
+
+### Signing Procedure
+
+To produce a signature, each signer `i` acts as follows.
+
+1. Produces a nonce pair `(k_i, R_i)` and a commitment (hash) `C_i` to `R_i`.
+   This is done with `secp256k1_musig_multisig_generate_nonce`. Sends the
+   commitment `C_i` to every other signer.
+2. Once at least `k` nonce commitments have been received from other signers,
+   signer `i` creates `n` `secp256k1_musig_signer_data` structures, one for
+   each signer including herself.
+   She initializes each with `secp256k1_musig_signer_data_initialize` which
+   takes a public key from all signers and a nonce commitment from present
+   signers. If `k = n`, the public key should be `Y_i` from step 2 of the key
+   setup; otherwise it should be `Z_i` from step 5 of key setup.
+3. She sends her actual public nonce `R_i` to every other signer.
+4. Upon receipt of another signer's public nonce, she calls `secp256k1_musig_set_nonce`
+   to update that signer's data structure. If the public nonce does not match
+   the precommitment, this function will fail and the signer will be considered
+   not to be present.
+5. Once `k` valid public nonces have been received, she can produce a partial
+   signature using `secp256k1_musig_partial_sign`. She sends this partial signature
+   to someone (or everyone) for aggregation.
+   [TODO] Things will work with more than `k` signers; if people disagree on the
+   set of signers this is a problem, so "just take the first `k`" is actually
+   likely to fail in the case of `>k`.
+6. Some participant receives all the partial signatures and combines them using
+   `secp256k1_musig_combine_partial_sigs`. The output of this function is a
+   complete signature. If the signature is invalid, at least one of the partial
+   signatures was invalid. The culprit may be identified using the function
+   `secp256k1_musig_partial_sig_verify` on the individual partial signatures.
+
```

Comment:
Each participant checks if other partial signatures are valid.  Whether the combined pubkey is correct is not checked, right?  I don't see any risk except wasting time in trying to produce final signature.  If would like to avoid this situation, can add checking combined pubkey the same time when checking nonce hash at the beginning of this procedure.